parquet/variants/primitive.go

```diff
@@ -0,0 +1,678 @@
+			}
+			dest.Set(reflect.ValueOf(float32(dv)))
+		case reflect.Float64, reflect.Interface:
+			dest.Set(reflect.ValueOf(dv))
+		default:
+			panic("unhandled float value")
+		}
+	case primitiveTimeNTZ, primitiveTimestampMicros, primitiveTimestampNTZMicros,
+		primitiveTimestampNanos, primitiveTimestampNTZNanos:
+		tsv, err := readUint(raw, offset+1, 8)
+		if err != nil {
+			return err
+		}
+
+		// Time can be decoded into either an int64 (the physical time), or into a time.Time struct.
+		// Anything else is invalid.
+		if kind == reflect.Int64 || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(int64(tsv)))
+		} else if kind == reflect.Uint64 {
+			dest.Set(reflect.ValueOf(tsv))
+		} else if dest.Type() == reflect.TypeOf(time.Time{}) {
+			var t time.Time
+			if prim == primitiveTimeNTZ {
+				// TimeNTZ for Variants is UTC=false (ie. local timezone) and in microseconds
+				t = time.Date(0, 0, 0, 0, 0, 0, 1000*int(tsv), time.Local)
+			} else {
+				if prim == primitiveTimestampMicros || prim == primitiveTimestampNTZMicros {
+					t = time.UnixMicro(int64(tsv))
+				} else {
+					sec := int64(tsv / 1e9)
+					nsec := int64(tsv % 1e9)
+					t = time.Unix(sec, nsec)
+				}
+				if prim == primitiveTimestampMicros || prim == primitiveTimestampNanos {
+					t = t.In(time.Local)
+				}
+			}
+			dest.Set(reflect.ValueOf(t))
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveString:
+		str, err := unmarshalString(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.String || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(str))
+		} else if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 {
+			dest.Set(reflect.ValueOf([]byte(str)))
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveBinary:
+		bytes, err := unmarshalBinary(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(bytes))
+		} else if kind == reflect.String {
+			dest.Set(reflect.ValueOf(string(bytes)))
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveUUID:
+		bytes, err := unmarshalUUID(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(bytes))
+		} else if kind == reflect.String {
+			dest.Set(reflect.ValueOf(string(bytes)))
+		} else {
+			return fmt.Errorf("cannot decode Variant UUID into dest %s", kind)
+		}
+	default:
+		return fmt.Errorf("unknown primitive: %s", prim)
+	}
+
+	return nil
+}
+
+func marshalNull(w io.Writer) int {
+	hdr, _ := primitiveHeader(primitiveNull)
+	w.Write([]byte{hdr})
+	return 1
+}
```

Comment:
the return for all of these should probably be (int, error) and just do `return w.Write(...)`